File: 1erdepartamental/c++/9.promedio_arreglo.cpp
// ============================================
// PROGRAMA: Estadísticas con Vector Dinámico
// DESCRIPCIÓN: Acepta cualquier cantidad de números
// ============================================

#include <iostream>
#include <vector>
using namespace std;


/**
 * Estadísticas sobre vector de tamaño variable
 * 
 * Más flexible que array fijo
 */
class EstadisticasArreglo {
private:
    vector<double> __datos;

public:
    EstadisticasArreglo(const vector<double>& arreglo)
        : __datos(arreglo) {}
    
    // Getter que retorna copia
    vector<double> getDatos() const {
        return __datos;
    }
    
    double suma() const {
        double total = 0;
        for (double n : __datos) {
            total += n;
        }
        return total;
    }
    
    double promedio() const {
        return suma() / __datos.size();
    }
    
    double maximo() const {
        double m = __datos[0];
        for (double n : __datos) {
            if (n > m) m = n;
        }
        return m;
    }
    
    double minimo() const {
        double m = __datos[0];
        for (double n : __datos) {
            if (n < m) m = n;
        }
        return m;
    }
    
    void agregar(double valor) {
        __datos.push_back(valor);
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    EstadisticasArreglo e({10, 25, 5, 30, 15});
    
    cout << "Datos iniciales: " << e.getDatos().size() << " elementos" << endl;
    cout << "Suma: " << e.suma() << endl;
    cout << "Promedio: " << e.promedio() << endl;
    cout << "Máximo: " << e.maximo() << endl;
    cout << "Mínimo: " << e.minimo() << endl << endl;
    
    e.agregar(50);
    cout << "Después de agregar 50:" << endl;
    cout << "Promedio: " << e.promedio() << endl;
    
    return 0;
}

File: 1erdepartamental/python/9.promedio_arreglo.py
# ============================================
# PROGRAMA: Estadísticas con Arreglo Dinámico
# DESCRIPCIÓN: Acepta cualquier cantidad de números
# ============================================


class EstadisticasArreglo:
    """
    Estadísticas sobre arreglo de tamaño variable.

    Más flexible que la versión de parámetros fijos.
    """

    def __init__(self, arreglo):
        self.__datos = arreglo.copy()

    @property
    def datos(self):
        return self.__datos.copy()

    def suma(self):
        total = 0
        for n in self.__datos:
            total += n
        return total

    def promedio(self):
        return self.suma() / len(self.__datos)

    def maximo(self):
        m = self.__datos[0]
        for n in self.__datos:
            if n > m:
                m = n
        return m

    def minimo(self):
        m = self.__datos[0]
        for n in self.__datos:
            if n < m:
                m = n
        return m

    def agregar(self, valor):
        """Ventaja sobre versión fija: puede crecer"""
        self.__datos.append(valor)


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    e = EstadisticasArreglo([10, 25, 5, 30, 15])

    print(f"Datos: {e.datos}")
    print(f"Suma: {e.suma()}")
    print(f"Promedio: {e.promedio()}")
    print(f"Máximo: {e.maximo()}")
    print(f"Mínimo: {e.minimo()}")

    e.agregar(50)
    print(f"\nDespués de agregar 50:")
    print(f"Datos: {e.datos}")
    print(f"Promedio: {e.promedio()}")
